modules/wallet/seed.go

```diff
@@ -39,121 +37,158 @@ type (
+func seedFileEncryptionKey(masterKey crypto.TwofishKey, sfuid SeedFileUID) crypto.TwofishKey {
+	return crypto.TwofishKey(crypto.HashAll(masterKey, seedModifier, sfuid))
 }
 
-// generateAndTrackKey will create key 'index' from seed 's', tracking the
-// public key. The secret key will be encrypted and stored.
-func (w *Wallet) generateAndTrackKey(masterKey crypto.TwofishKey, s seed, seedFilename string, index uint64) error {
-	// Generate the key.
-	entropy := crypto.HashAll(s, index)
-	sk, pk := crypto.DeterministicSignatureKeys(entropy)
-
-	// Fetch the unlock hash.
-	unlockHash := types.UnlockConditions{
+// generateUnlockConditions provides the unlock conditions that would be
+// automatically generated from the input public key.
+func generateUnlockConditions(pk crypto.PublicKey) types.UnlockConditions {
+	return types.UnlockConditions{
 		PublicKeys: []types.SiaPublicKey{{
 			Algorithm: types.SignatureEd25519,
 			Key:       pk[:],
 		}},
 		SignaturesRequired: 1,
-	}.UnlockHash()
+	}
+}
+
+// generateSpendableKey creates the keys and unlock conditions a given index of a
+// seed.
+func generateSpendableKey(seed modules.Seed, index uint64) spendableKey {
+	// Generate the keys and unlock conditions.
+	entropy := crypto.HashAll(seed, index)
+	sk, pk := crypto.DeterministicSignatureKeys(entropy)
+	return spendableKey{
+		unlockConditions: generateUnlockConditions(pk),
+		secretKeys:       []crypto.SecretKey{sk},
+	}
+}
 
-	// Encrypt the secret key.
-	skek := signatureKeyEncryptionKey(masterKey, seedFilename, index)
-	encryptedSignatureKey, err := skek.EncryptBytes(sk[:])
+// decryptSeedFile decrypts a seed file using the encryption key.
+func decryptSeedFile(masterKey crypto.TwofishKey, sf SeedFile) (seed modules.Seed, err error) {
+	// Verify that the provided master key is the correct key.
+	decryptionKey := seedFileEncryptionKey(masterKey, sf.SeedFileUID)
+	expectedDecryptedVerification := make([]byte, 32)
+	decryptedVerification, err := decryptionKey.DecryptBytes(sf.EncryptionVerification)
 	if err != nil {
-		return err
+		return seed, err
+	}
+	if !bytes.Equal(expectedDecryptedVerification, decryptedVerification) {
+		return seed, errBadEncryptionKey
 	}
 
-	// Add the key to the set of tracked keys.
-	w.generatedKeys[unlockHash] = generatedSignatureKey{index: index, encryptedKey: encryptedSignatureKey}
-	w.trackedKeys[unlockHash] = struct{}{}
-	return nil
+	// Decrypt and return the seed.
+	plainSeed, err := decryptionKey.DecryptBytes(sf.Seed)
+	if err != nil {
+		return seed, err
+	}
+	copy(seed[:], plainSeed)
+	return seed, nil
+}
+
+// generateAndTrackKey tracks a key of a given index from a given seed.
+func (w *Wallet) generateAndTrackKey(seed modules.Seed, index uint64) {
+	// Generate the key and check it is new to the wallet.
+	spendableKey := generateSpendableKey(seed, index)
+	w.keys[spendableKey.unlockConditions.UnlockHash()] = spendableKey
```

Comment:
I would probably inline this function unless you think it's going to be called in places other than `integrateSeed`. As it stands, the meaning of "index" is a big confusing.